library/red-black/map.lisp

```diff
@@ -0,0 +1,172 @@
+(coalton-library/utils:defstdlib-package :coalton-library/red-black/map
+    (:use
+     :coalton
+     :coalton-library/classes
+     :coalton-library/tuple
+     :coalton-library/functions)
+  (:local-nicknames (#:tree :coalton-library/red-black/tree)
+                    (#:iter :coalton-library/iterator))
+  (:shadow #:Map #:empty)
+  (:export
+   #:Map
+   #:empty
+   #:lookup
+   #:insert
+   #:replace
+   #:replace-or-insert #:insert-or-replace
+   #:remove
+   #:keys
+   #:values
+   #:entries
+   #:collect!
+   #:update))
+(cl:in-package :coalton-library/red-black/map)
+
+(coalton-toplevel
+  (define-type (MapPair :key :value)
+    (MapPair :key :value)
+
+    ;; hack: when searching for an entry, we need a thing that holds just a key, but is of type `MapPair :key
+    ;; :any', to pass to `tree:lookup' or `tree:remove'. A `Map' will never hold a `JustKey'.
+    (JustKey :key))
+
+  (define (key pair)
+    (match pair
+      ((MapPair k _) k)
+      ((JustKey k) k)))
+
+  (define (value pair)
+    (match pair
+      ((MapPair _ v) v)
+      (_ (error "Misuse of `JustKey': Cannot read the Value of a JustKey"))))
+
+  (define-instance (Into (MapPair :key :value) (Tuple :key :value))
+    (define (into pair)
+      (Tuple (key pair) (value pair))))
+
+  (define-instance ((Eq :key) => (Eq (MapPair :key :value)))
+    (define (== left right)
+      (== (key left) (key right))))
+
+  (define-instance ((Ord :key) => (Ord (MapPair :key :value)))
+    (define (<=> left right)
+      (<=> (key left) (key right))))
+
+  (repr :transparent)
+  (define-type (Map :key :value)
```

Comment:
I think map should be able to implement semigroup, monoid, functor, foldable, and traversable.

Reply:
I'm assuming you want `merge` (name pending) as the Semigroup operator.

`Map` with `merge` is not a lawful `Monoid`, because `Map` is not a multiset. If `x` contains a mapping `a -> x'` and `y` contains a mapping `a -> y'`, then `x <> y` will map `a -> x'` whereas `y <> x` will map `a -> y'`. Or possibly vice versa.

That said, `Semigroup`, `Functor`, `Foldable` and `Traversable` are all easy.